tests/unit/auth/test_signers.py

```diff
@@ -516,8 +591,129 @@ def test_presign_with_security_token(self):
         auth = botocore.auth.S3SigV4QueryAuth(
             self.credentials, self.service_name, self.region_name, expires=60)
         request = AWSRequest()
+        request.method = 'GET'
         request.url = 'https://ec2.us-east-1.amazonaws.com/'
         auth.add_auth(request)
         query_string = self.get_parsed_query_string(request)
         self.assertEqual(
             query_string['X-Amz-Security-Token'], 'security-token')
+
+
+class BaseS3PresignPostTest(unittest.TestCase):
+    def setUp(self):
+        self.access_key = 'access_key'
+        self.secret_key = 'secret_key'
+        self.credentials = botocore.credentials.Credentials(
+            self.access_key, self.secret_key)
+
+        self.service_name = 'myservice'
+        self.region_name = 'myregion'
+
+        self.bucket = 'mybucket'
+        self.key = 'mykey'
+        self.policy = {
+            "expiration": "2007-12-01T12:00:00.000Z",
+            "conditions": [
+                {"acl": "public-read"},
+                {"bucket": self.bucket},
+                ["starts-with", "$key", self.key],
+            ]
+        }
+        self.fields = {
+            'key': self.key,
+            'acl': 'public-read',
+        }
+
+        self.request = AWSRequest()
+        self.request.url = 'https://s3.amazonaws.com/%s' % self.bucket
+        self.request.method = 'POST'
+
+        self.request.context['s3-presign-post-fields'] = self.fields
+        self.request.context['s3-presign-post-policy'] = self.policy
+
+class TestS3SigV2Post(BaseS3PresignPostTest):
+    def setUp(self):
+        super(TestS3SigV2Post, self).setUp()
+        self.auth = botocore.auth.HmacV1PostAuth(self.credentials)
+
+        self.current_epoch_time = 1427427247.465591
+        self.time_patch = mock.patch('time.time')
+        self.time_mock = self.time_patch.start()
+        self.time_mock.return_value = self.current_epoch_time
+
+    def tearDown(self):
+        self.time_patch.stop()
+
+    def test_presign_post(self):
+        self.auth.add_auth(self.request)
+        result_fields = self.request.context['s3-presign-post-fields']
+        self.assertEqual(
+            result_fields['AWSAccessKeyId'], self.credentials.access_key)
+
+        result_policy = json.loads(base64.b64decode(
+            result_fields['policy']).decode('utf-8'))
+        self.assertEqual(result_policy['expiration'],
+                         '2007-12-01T12:00:00.000Z')
+        self.assertEqual(
+            result_policy['conditions'],
+            [{"acl": "public-read"},
+             {"bucket": "mybucket"},
+             ["starts-with", "$key", "mykey"]])
+        self.assertIn('signature', result_fields)
```

Comment:
I did not assert do assert equal because we dump a dictionary to json and there is no guarantee that gets dumped the same each time unless you order the json elements, which is not needed functionality-wise.